### ISSUE TITLE
Is Web the future of Urbit console? (Or "70' called and want their VT100 terminal back")

### ISSUE THREAD

**Issue by mikolajpp** [urbit/arvo#601]
Today I was thinking about the sorry state of Urbit console. No ncurses. All line oriented. 
Close to nil interactivity save for line-input.

Quite possible we won't get ncurses support in the near future. 
I want to be able to:
1. See clay merge/sync progress status.
2. Console apps could offer much more interesting UI
and so on.

On the other hand, I realized ncurses is not a solution. I mean, we are in the current year ;-)
I believe we should probably completely abandon the idea of ever baking in some support for ncurses into Vere.

Here is why:
Console can be made superior in every way in the browser.
"But what about my embedded system" objection, I believe, is outdated. 
Just login to your urbit via an open port using laptop/smartphone/anything, which you will eventually do anyway. For example routers have long adhered to the idea of "web interaction output" being guaranteed to exist.

I think there is no reason to put much more work into terminal console than already put. For rare diagnostic reasons, the current console is more than enough. We should probably treat it (at the risk of me being dragged into the street by bearded Unix folks. Meanwhile I belong to shaven Unix folks ;-)) the way we treat `dmesg` in linux. Handy logging tool with some degree of interactivity, which nonetheless at this stage of Urbit development is a primary means of interaction. 

To prove this point, I will be implementing a new web console using tools i have developed for Plato (don't have a name yet. Want to keep dojo?) as a proof of concept meant to become a default way of interacting with urbit. (One PR win will be the typing lag goes away)

The part that needs some thinking: how to actually enable this new type of "interaction output".

1. Should apps be forced to use a new library, possibly modified dojo/sole?
2. Should it be text based? Unless we are bent on enhancing unix terminal at some later 
    date, I think we could just engineer new library which would operate on general 
    "interactive data". If the target console would be web based, it would get transformed 
    into json. If at some later date a bearded individual would enable ncurses support, it will be up to him to handle that. 
3. Some other ingenious idea?

I will need to dig deeper into current working of Urbit console handling (and look at the terminal color implementation which is hanging somewhere), but no matter what I think this is definitely doable short-term.

**Comment by mikolajpp** [comment]
I think there could be a sort of category of "widgets" in addition for rich output. 
For example, a lot of apps will want to display progress. 

They would pass some structure like
'{ widget: progress-bar
    args: {
               progress: 0.9
               status: 'uploading your wallet'
              }
   default: "[90%] Uploading wallet"
}' 

If there would be no widget on the console, it would just print the fallback line, and so on. 
Lots of cool possibilities. 

The ncurses model where each app reimplements their own interface using basic elements such as "windows" and other horrid stuff should be forgotten for the good of us all ;-)

**Comment by cgyarvin** [comment]
There is much to be said for rich output.  Another example would be: when I ls an image folder, I should see thumbnails.

I think in most cases it should be possible to scale back to vt100.  But it does become confining at a certain point.  And you’re right, value is not that great.

We’ve actually thought harder about another issue: autocomplete / autohelp.  This area is dynamite for a discoverable command line...

**Comment by galenwp** [comment]
We worked a good bit on redesigning the console experience, but ultimately changed gears for the near term. 

The idea was to 1) make it possible to run Urbit as a daemon 2) split the console into a client and server and 3) upgrade the ui to be much closer to screen.

We came up with a number of ways of hacking around using ncurses in service of 3). I think I even have plain-text layouts somewhere that outlined how we were going to do this.

The console work is on a branch somewhere and be taken over. It'd be interesting to have you look at it. @keatondunsford can probably help dig it up. He also helped with some of the development and could provide some color here. @ohAitch knows quite a bit about this too.
